lib/db/schema/patch-065-066.sql

```diff
@@ -0,0 +1,83 @@
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+-- Add an index on sessionTokens::createdAt to fix pruning filesort.
+CREATE INDEX `sessionTokens_createdAt`
+ON `sessionTokens` (`createdAt`)
+ALGORITHM=INPLACE
+LOCK=NONE;
+
+-- Used to prevent session token pruning from doing a full table scan
+-- as it proceeds further and further through the (very long) backlog
+-- of pruning candidates.
+INSERT INTO dbMetadata (name, value)
+VALUES ('sessionTokensPrunedUntil', 0);
+
+-- Update the prune stored procedure with assorted jiggery-pokery to
+-- make session token pruning faster.
+CREATE PROCEDURE `prune_5` (IN `olderThan` BIGINT UNSIGNED)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  SELECT @lockAcquired := GET_LOCK('fxa-auth-server.prune-lock', 3);
+
+  IF @lockAcquired THEN
+    DELETE FROM accountResetTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM passwordForgotTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM passwordChangeTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM unblockCodes WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM signinCodes WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+
+    START TRANSACTION;
+
+    -- Jiggery-pokery #1: Find out how far we got on previous iterations.
+    SELECT @newerThan := value FROM dbMetadata WHERE name = 'sessionTokensPrunedUntil';
+
+    -- Jiggery-pokery #2: Store the pruned session tokens in a temporary table
+    --                    ahead of time, so we can refer to createdAt post-facto.
+    CREATE TEMPORARY TABLE prunedSessionTokens (
+      tokenId BINARY(32) PRIMARY KEY,
+      createdAt BIGINT UNSIGNED NOT NULL,
+      INDEX prunedSessionTokens_createdAt (createdAt)
+    ) AS
+      SELECT tokenId, createdAt FROM sessionTokens
+      WHERE createdAt >= @newerThan AND createdAt < olderThan
+      AND NOT EXISTS (
+        SELECT sessionTokenId FROM devices
+        WHERE sessionTokenId = sessionTokens.tokenId
+      )
+      ORDER BY createdAt
+      LIMIT 10000;
+
+    DELETE FROM sessionTokens
+    WHERE tokenId IN (
+      SELECT tokenId FROM prunedSessionTokens
+    );
+
+    DELETE ut
+    FROM unverifiedTokens AS ut
+    LEFT JOIN sessionTokens AS st
+      ON ut.tokenId = st.tokenId
+    WHERE st.tokenId IS NULL;
```

Comment:
Is this also a full table scan of the `unverifiedTokens` table?  I wonder if we can do it through a join as part of the above `DELETE FROM sessionTokens` query, i.e. `DELETE sessionTokens, unverifiedTokens FROM <the session-token deletion stuff> INNER JOIN unverifiedTokens`

Something else to check here: is `unverifiedTokens.tokenId` always a reference to `sessionTokens`?  I recall that `keyFetchTokens` could also be in verified or unverified state.

Reply:
> Is this also a full table scan of the `unverifiedTokens` table?

Oh dear, it is. I'll sort it out.

> I wonder if we can do it through a join...

This is covered in in the comments to #275 but unfortunately a delete with `INNER JOIN` doesn't work for us here. As per [the docs](https://dev.mysql.com/doc/refman/5.7/en/delete.html#idm140189111263328):

> You cannot use `ORDER BY` or `LIMIT` in a multiple-table `DELETE`.

It was literally the first thing I tried for #275, the query fails with a syntax error.

> Something else to check here: is `unverifiedTokens.tokenId` always a reference to `sessionTokens`?  I recall that `keyFetchTokens` could also be in verified or unverified state.

Well remembered! I forgot all about that. This gets messier and messier...

Reply:
> Is this also a full table scan of the `unverifiedTokens` table?

Way out of my depth here, but I assume that is what the `ALL` means in the `type` column returned by `EXPLAIN`?

```
mysql> explain select * from unverifiedTokens as ut left join sessionTokens as st
    -> on ut.tokenId = st.tokenId where st.tokenId is null;
+----+-------------+-------+------------+--------+---------------+---------+---------+----------------+------+----------+-------------------------+
| id | select_type | table | partitions | type   | possible_keys | key     | key_len | ref            | rows | filtered | Extra                   |
+----+-------------+-------+------------+--------+---------------+---------+---------+----------------+------+----------+-------------------------+
|  1 | SIMPLE      | ut    | NULL       | ALL    | NULL          | NULL    | NULL    | NULL           |    1 |   100.00 | NULL                    |
|  1 | SIMPLE      | st    | NULL       | eq_ref | PRIMARY       | PRIMARY | 32      | fxa.ut.tokenId |    1 |   100.00 | Using where; Not exists |
+----+-------------+-------+------------+--------+---------------+---------+---------+----------------+------+----------+-------------------------+
2 rows in set, 1 warning (0.00 sec)
```

> Something else to check here: is `unverifiedTokens.tokenId` always a reference to `sessionTokens`?  I recall that `keyFetchTokens` could also be in verified or unverified state.

The obvious (to me) way to fix this seems like adding another `LEFT JOIN` against `keyFetchTokens`, but then we'd have yet another full table scan. In hindsight, a `tokenType` column may have been useful at this point. Any other suggestions?

Reply:
(also, I only just realised I should be joining on `uid` and `tokenId` here, fixed in ef04618)

Reply:
> Something else to check here: is `unverifiedTokens.tokenId` always a reference to `sessionTokens`?  I recall that `keyFetchTokens` could also be in verified or unverified state.

Fix + test for the `keyFetchTokens` problem is in 20e3143.